### PR TITLE
=sha #20315 make sharding snapshotting based on seqNr, not counter

### LINKED ISSUE #20315: Shard should snapshot every [seqNr % when] and not [localCounter %when]
Sharding as a local variable `persistCount: Int` that it uses to decide when to snapshot,
by default we snapshot every 1000 events.

This counter is updated in saveSnapshotWhenNeeded() which is called when shard regions terminate or are allocated (when sending a message to a region that is not allocated yet, it does not seem to be updated during replays).

It could happen that the coordinator fails when it was around 900th event, and then starts again, goes to 900th again etc, without ever snapshotting. So we end up with more events to be replayed than one might expect from the snapshot setting.

### PR DESCRIPTION
**Pull request by ktoso**
Resolves #20315

### CODE REVIEW

**Comment by ktoso** [comment]
Please review @patriknw

**Comment by ktoso** [comment]
I'll add a test for this specifically in a bit

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 1]
```diff
@@ -334,8 +332,7 @@ private[akka] class PersistentShard(
   }
 
   def saveSnapshotWhenNeeded(): Unit = {
-    persistCount += 1
-    if (persistCount % snapshotAfter == 0) {
+    if (lastSequenceNr % snapshotAfter == 0) {
```
Comment (viktorklang) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: Wow, net negative, nice!

Comment (ktoso) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala]: Yeah, finally some `-` points :wink:

**Comment by patriknw** [comment]
there is a corresponding counter in `ShardCoordinator`

**Comment by patriknw** [comment]
Looks like this can be easily completed, there is a corresponding counter in ShardCoordinator. Something for Monday, @ktoso

**Comment by ktoso** [comment]
Addressed in `PersistentShardCoordinator`, rebased and pushed 👍

**Comment by patriknw** [comment]
LGTM, thanks

**Comment by ktoso** [comment]
Failure was MiMa, added filters finally

**Pull request merged**